File: control-plane/src/main/java/com/kratisai/controlplane/api/wsdto/ActivityKind.java
package com.kratisai.controlplane.api.wsdto;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.Locale;
import java.util.Optional;

/**
 * Closed ACP tool-kind set plus WRITE, the sidecar alias for fs/write_text_file.
 * Backs both activity detail kinds and HITL permission/rule matching, so the
 * vocabulary is defined exactly once.
 */
public enum ActivityKind {
    READ("read"),
    EDIT("edit"),
    WRITE("write"),
    DELETE("delete"),
    MOVE("move"),
    SEARCH("search"),
    EXECUTE("execute"),
    THINK("think"),
    FETCH("fetch"),
    SWITCH_MODE("switch_mode"),
    OTHER("other");

    private final String value;

    ActivityKind(String value) {
        this.value = value;
    }

    @JsonValue
    public String getValue() {
        return value;
    }

    /** Lenient wire-value lookup: empty when the value is absent or unknown. */
    public static Optional<ActivityKind> fromWireValue(String value) {
        if (value == null) {
            return Optional.empty();
        }
        String normalized = value.trim();
        for (ActivityKind kind : values()) {
            if (kind.value.equalsIgnoreCase(normalized) || kind.name().equalsIgnoreCase(normalized)) {
                return Optional.of(kind);
            }
        }
        return Optional.empty();
    }

    @JsonCreator
    public static ActivityKind fromString(String val) {
        return fromWireValue(val).orElseThrow(() -> new IllegalArgumentException("Unknown activity kind: " + val));
    }

    /** Wire value for a possibly-absent tool kind, defaulting to EXECUTE. */
    public static String effectiveWireValue(String toolKind) {
        if (toolKind == null || toolKind.isBlank()) {
            return EXECUTE.value;
        }
        return toolKind.trim().toLowerCase(Locale.ROOT);
    }

    public static boolean isCommandLike(String toolKind) {
        return EXECUTE.value.equals(effectiveWireValue(toolKind));
    }
}

File: web/src/types/activity-kind.ts
/**
 * Single source of truth for the tool-kind vocabulary shared by activity
 * records, the protocol zod enum, and HITL permission narrowing.
 *
 * The values mirror the ACP ToolKind set (plus `write`, Kratis's alias for
 * fs/write_text_file) and the control-plane `ActivityKind` enum.
 */
export const ACTIVITY_KINDS = [
  'read',
  'edit',
  'write',
  'delete',
  'move',
  'search',
  'execute',
  'think',
  'fetch',
  'switch_mode',
  'other',
] as const

export type ActivityKind = (typeof ACTIVITY_KINDS)[number]

const ACTIVITY_KIND_SET: ReadonlySet<string> = new Set(ACTIVITY_KINDS)

/**
 * Narrows an unconstrained wire `toolKind` to a known ActivityKind.
 * ACP permission requests carry `toolKind` as a plain string, so unknown or
 * absent values degrade to `other` rather than dropping the activity.
 */
export function toActivityKind(toolKind: string | undefined): ActivityKind {
  const candidate = toolKind?.toLowerCase() ?? ''
  return ACTIVITY_KIND_SET.has(candidate) ? (candidate as ActivityKind) : 'other'
}
